lsp/nls/src/background.rs

```diff
@@ -15,305 +14,236 @@ use nickel_lang_core::{
         }
-        evals = dedup;
-
-        if let Some(uri) = evals.pop() {
-            let Ok(path) = uri_to_path(&uri) else {
-                warn!("skipping invalid uri {uri}");
-                continue;
-            };
-
-            if let Some(file_id) = world.cache.id_of(&SourcePath::Path(path.clone())) {
-                response_tx
-                    .send(Response::Starting { uri: uri.clone() })
-                    .ok()?;
-
-                let mut diagnostics = world.parse_and_typecheck(file_id);
-
-                // Evaluation diagnostics (but only if there were no parse/type errors).
-                if diagnostics.is_empty() {
-                    // TODO: avoid cloning the cache.
-                    let mut vm =
-                        VirtualMachine::<_, CacheImpl>::new(world.cache.clone(), std::io::stderr());
-                    // We've already checked that parsing and typechecking are successful, so we
-                    // don't expect further errors.
-                    let rt = vm.prepare_eval(file_id).unwrap();
-                    let errors = vm.eval_permissive(rt);
-                    diagnostics.extend(
-                        errors
-                            .into_iter()
-                            .filter(|e| {
-                                !matches!(
-                                    e,
-                                    nickel_lang_core::error::EvalError::MissingFieldDef { .. }
-                                )
-                            })
-                            .flat_map(|e| world.lsp_diagnostics(file_id, e)),
-                    );
-                }
 
-                // If there's been an update to the file, don't send back a stale response.
-                cmds.extend(cmd_rx.try_recv());
-                if !cmds.iter().any(|cmd| match cmd {
-                    Command::UpdateFile { uri, .. } => {
-                        uri_to_path(uri).map_or(false, |p| p == path)
-                    }
-                    _ => false,
-                }) {
-                    response_tx
-                        .send(Response::Diagnostics(Diagnostics { path, diagnostics }))
-                        .ok()?;
-                }
-            }
-        }
+        let diagnostics = Diagnostics { path, diagnostics };
 
-        drain_ready(&cmd_rx, &mut cmds);
-        if cmds.is_empty() && evals.is_empty() {
-            // Wait for a command to be available.
-            cmds.push(cmd_rx.recv().ok()?);
-        }
+        // If this fails, the main process has already exited. No need for a loud error in that case.
+        let _ = bincode::serialize_into(std::io::stdout().lock(), &diagnostics);
     }
+
+    Ok(())
 }
 
 struct SupervisorState {
     cmd_rx: Receiver<Command>,
     response_tx: Sender<Diagnostics>,
-    child: Child,
-    cmd_tx: IpcSender<Command>,
-    response_rx: Receiver<Response>,
 
     contents: HashMap<Url, String>,
+    deps: HashMap<Url, Vec<Url>>,
```

Comment:
Is this related to running the process in the background? Why do we need to keep deps now?

Reply:
Because we spin up a new process for each job, we need to transfer over all the file contents (because the LSP's in-memory contents might be different to what's on disk). The dependency tracking allows us to send over only the file contents that the current eval depends on.